### PR TITLE
allow MD Spark to write metrics to hdfs and gcs

### LINKED ISSUE #1025: Can't write Metrics files to gcs
It looks like it only works with NFS.
It's easy to reproduce using MarkDuplicatesSpark.

### LINKED ISSUE #1026: Can't write Metrics files to HDFS
Writing metrics only works with NFS.
This is easy to reproduce with MakeDuplicatesSpark.

### PR DESCRIPTION
**Pull request by lbergelson**
change MarkDuplicatesSpark and MarkDuplicatesSparkUtils so that metrics can be written to hdfs:// and gs:// paths.

Fixes #1025 
Fixes #1026

[Review on Reviewable](https://reviewable.io/reviews/broadinstitute/gatk/1044)

### CODE REVIEW

**Comment by lbergelson** [comment]
@davidaadams updated with a test as well as some new convinience test methods (which themselves don't yet have tests...)

File: src/main/java/org/broadinstitute/hellbender/utils/gcs/BucketUtils.java [hunk 1]
```diff
@@ -175,6 +181,50 @@ public static void deleteFile(String pathToDelete, PipelineOptions popts) throws
...
+     *
+     */
+    public static String getTempFilePath(String prefix, String extension, AuthHolder authHolder){
+        if (BucketUtils.isCloudStorageUrl(prefix) || (BucketUtils.isHadoopUrl(prefix))){
+            final String path = randomRemotePath(prefix, "", extension);
+            deleteOnExit(path, authHolder);
+            deleteOnExit(path + Tribble.STANDARD_INDEX_EXTENSION, authHolder);
+            deleteOnExit(path + TabixUtils.STANDARD_INDEX_EXTENSION, authHolder);
+            deleteOnExit(path + ".bai", authHolder);
+            deleteOnExit(path.replaceAll(extension + "$", ".bai"), authHolder);
```
Comment (davidadamsphd) [inline src/main/java/org/broadinstitute/hellbender/utils/gcs/BucketUtils.java]: What's happening on this line? (Please put the answer in a comment).

**Comment by davidadamsphd** [comment]
Not sure why this is failing (null pointer exception in one of your tests), but it's pretty much what I wanted to see. Once you fix the test and address my one question, it's ready to be merged.

**Comment by lbergelson** [comment]
I'm also confused about why it's failing.  It's something to do with how the `@BeforeClass` annotation is working.  (working locally, not working on traivs :/)

**Pull request merged**